### PR TITLE
Downgrade part2

### PR DESCRIPTION
**Pull request by YoyinZyc**

### CODE REVIEW

File: Documentation/dev-guide/api_reference_v3.md [hunk 1]
```diff
@@ -461,7 +461,7 @@ Empty field.
 | ----- | ----------- | ---- |
 | header |  | ResponseHeader |
 | version | version is the current cluster version. | string |
-| status | status is how the version of the servers in the cluster like. | string |
+| status | status is the situation of each server's version in the cluster | string |
```
Comment (wenjiaswe) [inline Documentation/dev-guide/api_reference_v3.md:464]: "Status" is not implemented yet, right? How about we add this when we implement it?

File: etcdctl/ctlv3/command/downgrade_command.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	dc.AddCommand(NewDowngradeCancelCommand())
+
+	return dc
+}
+
+// NewDowngradeValidateCommand returns the cobra command for "downgrade validate"
+func NewDowngradeValidateCommand() *cobra.Command {
+	cc := &cobra.Command{
+		Use:   "validate <targetVersion>",
+		Short: "Validate the downgrade capability against target version",
```
Comment (wenjiaswe) [inline etcdctl/ctlv3/command/downgrade_command.go]: "Validate if the cluster can be downgraded to target version" or something? Capability has specific meaning in etcdserver/api/capability.go, let's try not get confusion with that.

File: etcdctl/ctlv3/command/printer_fields.go [hunk 3]
```diff
@@ -226,3 +226,21 @@ func (p *fieldsPrinter) UserRevokeRole(user string, role string, r v3.AuthUserRe
...
+	fmt.Println("Validate succeeded")
+	fmt.Println(`"Current cluster version" :`, r.Version)
+}
+
+func (p *fieldsPrinter) DowngradeEnable(r v3.DowngradeResponse) {
+	p.hdr(r.Header)
+	fmt.Println("The cluster is available to downgrade. Please wait until all servers are ready.")
+	fmt.Println(`"Current cluster version" :`, r.Version)
+}
+
```
Comment (wenjiaswe) [inline etcdctl/ctlv3/command/printer_fields.go:241]: Do we want shorter response here?

File: etcdserver/api/etcdhttp/peer.go [hunk 4]
```diff
@@ -34,23 +34,32 @@ import (
...
 	return newPeerHandler(lg, s, s.RaftHandler(), s.LeaseHandler())
 }
 
-func newPeerHandler(lg *zap.Logger, s etcdserver.Server, raftHandler http.Handler, leaseHandler http.Handler) http.Handler {
+func newPeerHandler(lg *zap.Logger, s etcdserver.ServerPeer, raftHandler http.Handler, leaseHandler http.Handler) http.Handler {
 	peerMembersHandler := newPeerMembersHandler(lg, s.Cluster())
 	peerMemberPromoteHandler := newPeerMemberPromoteHandler(lg, s)
+	// downgradeStatusHandler handles the request to get the downgrade status in a cluster;
+	// only leader will return its downgrade status
+	downgradeStatusHandler := newDowngradeStatusHandler(lg, s)
```
Comment (wenjiaswe) [inline etcdserver/api/etcdhttp/peer.go]: Remove status related?

File: etcdserver/api/etcdhttp/peer.go [hunk 5]
```diff
@@ -85,6 +94,32 @@ type peerMemberPromoteHandler struct {
...
+		server:  s,
+	}
+}
+
+type downgradeStatusHandler struct {
+	lg      *zap.Logger
+	cluster api.Cluster
+	server  etcdserver.ServerPeer
+}
+
```
Comment (wenjiaswe) [inline etcdserver/api/etcdhttp/peer.go]: Status related.

File: etcdserver/api/etcdhttp/peer.go [hunk 6]
```diff
@@ -157,3 +192,55 @@ func (h *peerMemberPromoteHandler) ServeHTTP(w http.ResponseWriter, r *http.Requ
 		}
 	}
 }
+
+func (h *downgradeStatusHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !allowMethod(w, r, "GET") {
```
Comment (wenjiaswe) [inline etcdserver/api/etcdhttp/peer.go]: status related.

File: etcdserver/api/membership/cluster.go [hunk 7]
```diff
@@ -60,14 +60,14 @@ type RaftCluster struct {
 	// removed id cannot be reused.
 	removed map[types.ID]bool
 
-	downgrade Downgrade
+	downgrade *Downgrade
 }
 
 type Downgrade struct {
-	// the target downgrade version, if the cluster is not under downgrade,
+	// TargetVersion is the target downgrade version, if the cluster is not under downgrade,
```
Comment (wenjiaswe) [inline etcdserver/api/membership/cluster.go]: s/under downgrade/downgrading?

File: etcdserver/api/membership/cluster.go [hunk 8]
```diff
@@ -796,44 +809,61 @@ func ValidateClusterAndAssignIDs(lg *zap.Logger, local *RaftCluster, existing *R
...
 			} else {
-				plog.Fatalf("cluster cannot be downgraded (current version: %s is lower than determined cluster version: %s).", version.Version, version.Cluster(cv.String()))
+				plog.Fatalf("invalid downgrade; server version is lower than determined cluster version (current version: %s, determined cluster version: %s).",
+					version.Version, version.Cluster(cv.String()))
 			}
 		}
 	}
 }
 
 func IsVersionChangable(cv *semver.Version, lv *semver.Version) bool {
```
Comment (wenjiaswe) [inline etcdserver/api/membership/cluster.go:871]: Sorry I probably missed the comments when I added this in, could you help to add?
// IsVersionChangable checks the two scenario when version is changable:
// 1. Downgrade: cluster version is 1 minor version higher than local version, cluster version should change
// 2. Cluster start: when not all members version are available,  cluster version is set to MinVersion (3.0), when
//    all members are at higher version (3.2), cluster version is lower than local version, cluster version should change

File: etcdserver/api/membership/cluster_test.go [hunk 9]
```diff
@@ -859,3 +880,221 @@ func TestIsReadyToRemoveVotingMember(t *testing.T) {
...
+	lv = &semver.Version{Major: lv.Major, Minor: lv.Minor}
+	oneMinorHigher := &semver.Version{Major: lv.Major, Minor: lv.Minor + 1}
+	oneMinorLower := &semver.Version{Major: lv.Major, Minor: lv.Minor - 1}
+	downgradeEnabledHigherVersion := &Downgrade{Enabled: true, TargetVersion: oneMinorHigher}
+	downgradeEnabledEqualVersion := &Downgrade{Enabled: true, TargetVersion: lv}
+	downgradeEnabledLowerVersion := &Downgrade{Enabled: true, TargetVersion: oneMinorLower}
+	downgradeDisabled := &Downgrade{Enabled: false}
+
+	tests := []struct {
+		currentVersion *semver.Version
```
Comment (wenjiaswe) [inline etcdserver/api/membership/cluster_test.go]: Let's add test name here so we know which one fails, something like "Downgrade enabled, target version match, should succeed."

File: etcdserver/api/membership/cluster_test.go [hunk 10]
```diff
@@ -859,3 +880,221 @@ func TestIsReadyToRemoveVotingMember(t *testing.T) {
...
+			oneMinorLower,
+			downgradeDisabled,
+			true,
+			"",
+		},
+		{
+			oneMinorHigher,
+			downgradeDisabled,
+			false,
+			"invalid downgrade; server version is lower than determined cluster version",
```
Comment (wenjiaswe) [inline etcdserver/api/membership/cluster_test.go:911]: determined cluster version -> downgrade target version?

File: etcdserver/cluster_util.go [hunk 11]
```diff
@@ -177,6 +177,93 @@ func getVersions(lg *zap.Logger, cl *membership.RaftCluster, local types.ID, rt
 	return vers
 }
 
+// isDowngradeEnabled returns the downgrade status in the given cluster.
+// If the local server or any other remote server has already enable downgrade, return true and target version.
```
Comment (wenjiaswe) [inline etcdserver/cluster_util.go]: s/enable/enabled

**Review by wenjiaswe — COMMENTED** [review]
Great job! Thanks! Have some nits that we already talked about offline.

@jingyih could you please help to take a look?

**Comment by jingyih** [comment]
Is the part 1 here?
https://github.com/wenjiaswe/etcd/pull/3

**Comment by YoyinZyc** [comment]
> Is the part 1 here?
> #3

yes

File: etcdserver/api/membership/cluster_test.go (hunk 10 above)
Comment (YoyinZyc) [reply etcdserver/api/membership/cluster_test.go:911]: It is should be `determined cluster version`. When downgrade is disabled, the `mustDetectDowngrade` function will compare local version with the determined cluster version.

File: etcdserver/api/membership/cluster_test.go (hunk 9 above)
Comment (YoyinZyc) [reply etcdserver/api/membership/cluster_test.go]: Added.

File: etcdserver/api/membership/cluster.go (hunk 8 above)
Comment (YoyinZyc) [reply etcdserver/api/membership/cluster.go:871]: Added

File: Documentation/dev-guide/api_reference_v3.md (hunk 1 above)
Comment (YoyinZyc) [reply Documentation/dev-guide/api_reference_v3.md:464]: sgtm. removed all status related

**Pull request merged**